### PR TITLE
Configure Spring Security, implement email validation, password valid…

### PR DESCRIPTION
**Pull request by Alkey**
…ation, implement CustomGlobalExceptionHandler

### CODE REVIEW

File: src/main/java/com/dev/cinema/validators/EmailValidator.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+import javax.validation.ConstraintValidatorContext;
+
+public class EmailValidator implements ConstraintValidator<ValidEmail, String> {
+    public static final Pattern VALID_EMAIL_ADDRESS_REGEX =
+            Pattern.compile("^[A-Z0-9._%+-]+@[A-Z0-9.-]+\\.[A-Z]{2,6}$",
+                    Pattern.CASE_INSENSITIVE);
+
+    @Override
+    public boolean isValid(String email, ConstraintValidatorContext constraint) {
+        Matcher matcher = VALID_EMAIL_ADDRESS_REGEX.matcher(email);
```
Comment (AnaLanda) [inline src/main/java/com/dev/cinema/validators/EmailValidator.java]: I think you should add a check for a `null` email here or in the DTO

File: src/main/java/com/dev/cinema/validators/PasswordValidator.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    @Override
+    public void initialize(ValidPassword constraintAnnotation) {
+        this.password = constraintAnnotation.password();
+        this.repeatPassword = constraintAnnotation.repeatPassword();
+    }
+
+    @Override
+    public boolean isValid(Object value, ConstraintValidatorContext constraintValidatorContext) {
+        Object dtoPassword = new BeanWrapperImpl(value)
+                .getPropertyValue(password);
```
Comment (AnaLanda) [inline src/main/java/com/dev/cinema/validators/PasswordValidator.java:21]: If the `password` is `null`, will `.getPropertyValue(password)` throw an NPE? If so, we need a null check before this method is called.

**Review by AnaLanda — COMMENTED** [review]
Looks good!

File: src/main/java/com/dev/cinema/config/SecurityConfig.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+                .password(getEncoder().encode("123"))
+                .roles("USER");
+    }
+
+    protected void configure(HttpSecurity http) throws Exception {
+        http
+                .authorizeRequests()
+                .anyRequest().authenticated()
+                .and()
+                .formLogin()
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/config/SecurityConfig.java:29]: .permitAll()?

File: src/main/java/com/dev/cinema/controllers/AuthenticationController.java [hunk 4]
```diff
@@ -1,14 +1,15 @@
...
 import com.dev.cinema.security.AuthenticationService;
+import javax.validation.Valid;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.web.bind.annotation.PostMapping;
 import org.springframework.web.bind.annotation.RequestBody;
 import org.springframework.web.bind.annotation.RestController;
 
 @RestController
-public class AuthenticationController {
+public class AuthenticationController extends CustomGlobalExceptionHandler {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/controllers/AuthenticationController.java]: why do you extend it from CustomGlobalExceptionHandler?

File: src/main/java/com/dev/cinema/controllers/MovieSessionController.java [hunk 5]
```diff
@@ -15,7 +16,7 @@
 import org.springframework.web.bind.annotation.RestController;
 
 @RestController
-public class MovieSessionController {
+public class MovieSessionController extends CustomGlobalExceptionHandler {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/controllers/MovieSessionController.java]: same

File: src/main/java/com/dev/cinema/controllers/OrderController.java [hunk 6]
```diff
@@ -9,15 +9,15 @@
...
 import org.springframework.web.bind.annotation.GetMapping;
 import org.springframework.web.bind.annotation.PostMapping;
 import org.springframework.web.bind.annotation.RequestMapping;
-import org.springframework.web.bind.annotation.RequestParam;
 import org.springframework.web.bind.annotation.RestController;
 
 @RestController
 @RequestMapping("/orders")
-public class OrderController {
+public class OrderController extends CustomGlobalExceptionHandler {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/controllers/OrderController.java]: same for all controllers

File: src/main/java/com/dev/cinema/model/dto/ShoppingCartRequestDto.java [hunk 7]
```diff
@@ -1,9 +1,12 @@
 package com.dev.cinema.model.dto;
 
+import javax.validation.constraints.NotNull;
 import lombok.Data;
 
 @Data
 public class ShoppingCartRequestDto {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/model/dto/ShoppingCartRequestDto.java]: let's remove this class

File: src/main/java/com/dev/cinema/validators/EmailValidator.java (hunk 1 above)
Comment (KseniiaMakarova) [reply src/main/java/com/dev/cinema/validators/EmailValidator.java]: let's add the null check for email here as well

File: src/main/java/com/dev/cinema/validators/PasswordValidator.java (hunk 2 above)
Comment (KseniiaMakarova) [reply src/main/java/com/dev/cinema/validators/PasswordValidator.java:21]: @AnaLanda `password` is the field name at the moment

File: src/main/java/com/dev/cinema/controllers/CustomGlobalExceptionHandler.java [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+import java.util.stream.Collectors;
+import org.springframework.context.support.DefaultMessageSourceResolvable;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.MethodArgumentNotValidException;
+import org.springframework.web.context.request.WebRequest;
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+
+public class CustomGlobalExceptionHandler extends ResponseEntityExceptionHandler {
```
Comment (KseniiaMakarova) [inline src/main/java/com/dev/cinema/controllers/CustomGlobalExceptionHandler.java:18]:
```suggestion
@ControllerAdvice
public class CustomGlobalExceptionHandler extends ResponseEntityExceptionHandler {
```

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**